src/values.rs

```diff
@@ -0,0 +1,242 @@
+            HashableValue::Int(v) => format!("const:int:{}", v),
+            HashableValue::Container(c) => c.desc(),
+        }
+    }
+
+    fn check_type(&self, ty: &HashableType) -> Result<(), ConstTypeError> {
+        if let HashableType::Container(Container::Alias(s)) = ty {
+            return Err(ConstTypeError::NoAliases(s.to_string()));
+        };
+        match self {
+            HashableValue::String(_) => {
+                if let HashableType::String = ty {
+                    return Ok(());
+                };
+            }
+            HashableValue::Int(value) => {
+                if let HashableType::Int(width) = ty {
+                    return check_int_fits_in_width(*value, *width).map_err(ConstTypeError::Int);
+                };
+            }
+            HashableValue::Container(vals) => {
+                if let HashableType::Container(c_ty) = ty {
+                    return vals.check_container(c_ty);
+                };
+            }
+        }
+        Err(ConstTypeError::ValueCheckFail(
+            ClassicType::Hashable(ty.clone()),
+            ConstValue::Hashable(self.clone()),
+        ))
+    }
+
+    fn container_error(
+        typ: Container<HashableType>,
+        vals: ContainerValue<HashableValue>,
+    ) -> ConstTypeError {
+        ConstTypeError::ValueCheckFail(
+            ClassicType::Hashable(HashableType::Container(typ)),
+            ConstValue::Hashable(HashableValue::Container(vals)),
+        )
+    }
+}
+
+/// A value that is a container of other values, e.g. a tuple or sum;
+/// thus, corresponding to [Container]. Note there is no member
+/// corresponding to [Container::Alias]; such types must have been
+/// resolved to concrete types in order to create instances (values).
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub enum ContainerValue<T> {
+    /// A [Container::Array] or [Container::Tuple] or [Container::List]
+    Sequence(Vec<T>),
+    /// A [Container::Map]
+    Map(Vec<(HashableValue, T)>), // TODO try to make this an actual map?
+    /// A [Container::Sum] - for any Sum type where this value meets
+    /// the type of the variant indicated by the tag
+    Sum(usize, Box<T>), // Tag and value
+    /// A value of a custom type defined by an extension/[Resource].
+    ///
+    /// [Resource]: crate::resource::Resource
+    // TODO replace this with CustomConst
+    Opaque(serde_yaml::Value),
+}
+
+impl<Elem: ValueOfType> ContainerValue<Elem> {
+    pub(crate) fn desc(&self) -> String {
+        match self {
+            ContainerValue::Sequence(vals) => {
+                let names: Vec<_> = vals.iter().map(ValueOfType::name).collect();
+                format!("const:seq:{{{}}}", names.join(", "))
+            }
+            ContainerValue::Map(_) => "a map".to_string(),
+            ContainerValue::Sum(tag, val) => format!("const:sum:{{tag:{tag}, val:{}}}", val.name()),
+            ContainerValue::Opaque(c) => format!("const:{:?}", c),
```

Comment:
worth saying opaque here?

Reply:
Used `const:custom:` and `const:yaml:` for the two different forms that we still have ATM